### PR TITLE
story #2105064: adapt configuration to support multi-workspaces for the same workspace configuration

### PR DESCRIPTION
**Pull request by pikachugb**
https://center.almoctane.com/ui/entity-navigation?p=1001/1002&entityType=work_item&id=2105064

(if confused about some nonexisting new code in this PR, try to check the coverage one aswell - https://github.com/MicroFocus/octane-jira-coverage-plugin/pull/31)

NOTE: some classes were just moved and not refactored when tho' they are not well written

- created a new configuration (v3) which has the same spaceConfiguration with a different workspaceConfiguration
- the new workspaceConfiguration has been refactored (besides making the octane workspace a set), splitting it into 2 parts: octane part (octaneConfigGrouping) and jira part (jiraConfigGrouping)
- added a new upgrader (upgrading configuration from v2 to v3 - for reference v2 represents the old configuration which is the multi-space configuration)
- change the validation of the space configurations to be unique only by their name (instead of name + octane location)
- create a multi-select dropdown instead of single-select for workspaces inside workspace dialog
image

- for the workspace configuration table, the columns 'Workspace Id' and 'Workspace Name' will be merged into a single column called 'Workspaces' which will get the data from the server in the format "_workspaceId - workspaceName_" (e.g. "1002 - default_workspace) -> the data is being passed through Outgoing objects
image

- the udf and octane entity types that can be used inside workspace dialog will be the common ones between all the workspaces selected in the new dropdown
- added a validation for the workspace configurations (both front and back) so 2 identical (same octaneworkspaces, same spaceConfigId, same Udf, same Jira projects+keys) cannot be added
image

still have to do in a separate US/def: add/edit license text

### CODE REVIEW

File: src/main/java/com/microfocus/octane/plugins/admin/ConfigResource.java [hunk 1]
```diff
@@ -120,27 +122,52 @@ public Response getWorkspaceConfigurationById(@PathParam("id") String id) {
...
     @Path("/workspaces/possible-jira-fields")
-    public Response getPossibleJiraFields(@QueryParam("space-config-id") String spaceConfigurationId, @QueryParam("workspace-id") long workspaceId) {
+    public Response getPossibleJiraFields(@QueryParam("space-config-id") String spaceConfigurationId, @QueryParam("workspace-ids") String workspaceIds) {
         if (!hasPermission()) {
             return Response.status(Status.FORBIDDEN).build();
         }
 
+        List<Long> workspaceIdsList;
+        try {
+            workspaceIdsList = Arrays.stream(workspaceIds.split(",")).map(Long::parseLong).collect(Collectors.toList());
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/admin/ConfigResource.java:158]: what happens if workspaceIds comes as null or invalid value?

File: src/main/java/com/microfocus/octane/plugins/admin/ConfigResource.java [hunk 2]
```diff
@@ -340,6 +368,13 @@ public Response testSpaceConfiguration(SpaceConfigurationOutgoing spaceConfigura
         }
     }
 
+    private boolean doesWorkspaceConfigurationAlreadyExists(WorkspaceConfiguration wc) {
+        return ConfigurationManager.getInstance().getWorkspaceConfigurations().stream()
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/admin/ConfigResource.java:372]: this is so hard to read. maybe properly name wsc and wc

File: src/main/java/com/microfocus/octane/plugins/configuration/ConfigurationManager.java [hunk 3]
```diff
@@ -182,10 +182,12 @@ public synchronized void clearConfiguration() {
 
     private ConfigurationCollection loadConfiguration() {
         PluginSettings settings = pluginSettingsFactory.createGlobalSettings();
-        String confStr = ((String) settings.get(CONFIGURATION_KEY_V2));
+        String confStr = readConfigurationForDataKey(settings, CONFIGURATION_KEY_V3);
         log.info("Configuration is loading...");
-        if (confStr == null) {//create initial configuration
-            configuration = tryConvertFromPreviousVersion(settings);
+
+        if (confStr == null) {
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/configuration/ConfigurationManager.java:188]: C: use isBlank from String utils. its safer cause it also checks if string is empty

File: src/main/java/com/microfocus/octane/plugins/configuration/ConfigurationManager.java [hunk 4]
```diff
@@ -204,53 +206,27 @@ private ConfigurationCollection loadConfiguration() {
...
-                        tempConfiguration.setProxy(configurationV1.getProxy());
-                        tempConfiguration.getProxy().setNonProxyHost("");
-                    }
-                }
-            } catch (Exception e) {
-                log.error("Failed tryConvertFromPreviousVersion : " + e.getMessage());
+    private ConfigurationCollection upgradeConfigurationFromPreviousVersions(PluginSettings settings) {
+        String confStrV2 = readConfigurationForDataKey(settings, CONFIGURATION_KEY_V2);
+
+        if (confStrV2 != null) {
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/configuration/ConfigurationManager.java:212]: same as above

File: src/main/java/com/microfocus/octane/plugins/configuration/v2/upgrader/UpgraderFromV1ToV2.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.UUID;
+
+import static com.microfocus.octane.plugins.configuration.ConfigurationManagerConstants.*;
+
+public class UpgraderFromV1ToV2 {
+
+    private static final Logger log = LoggerFactory.getLogger(UpgraderFromV1ToV2.class);
+
+    public static void upgradeConfigurationFromV1ToV2(PluginSettings settings, PluginSettingsFactory pluginSettingsFactory) {
+        ConfigurationCollectionV2 tempConfiguration = null;
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/configuration/v2/upgrader/UpgraderFromV1ToV2.java:28]: shouldn't you initialize it with new ConfigurationCollectionV2()?

File: src/main/java/com/microfocus/octane/plugins/configuration/v2/upgrader/UpgraderFromV1ToV2.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        String confStrV1 = ((String) settings.get(CONFIGURATION_KEY_V1));
+
+        if (confStrV1 != null) {
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/configuration/v2/upgrader/UpgraderFromV1ToV2.java:31]: same isBlank

File: src/main/java/com/microfocus/octane/plugins/configuration/v3/upgrader/UpgraderFromV2ToV3.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class UpgraderFromV2ToV3 {
+
+    private static final Logger log = LoggerFactory.getLogger(UpgraderFromV2ToV3.class);
+
+    public static ConfigurationCollection upgradeConfigurationFromV2ToV3(String confStrV2) {
+        ConfigurationCollection tempConfiguration = null;
```
Comment (timisaratean) [inline src/main/java/com/microfocus/octane/plugins/configuration/v3/upgrader/UpgraderFromV2ToV3.java:19]: same q with new as above

**Review by timisaratean — COMMENTED** [review]
as an  overall comment, find better names for variables and consts
its hard to follow the code

File: src/main/java/com/microfocus/octane/plugins/configuration/ConfigurationManager.java (hunk 3 above)
Comment (pikachugb) [reply src/main/java/com/microfocus/octane/plugins/configuration/ConfigurationManager.java:188]: in this case, I don't really care about empty string, I want to check if previous configuration exists at all in the jira server
if it's empty string it's a different case - it existed, but it was empty

File: src/main/java/com/microfocus/octane/plugins/configuration/v2/upgrader/UpgraderFromV1ToV2.java (hunk 5 above)
Comment (pikachugb) [reply src/main/java/com/microfocus/octane/plugins/configuration/v2/upgrader/UpgraderFromV1ToV2.java:28]: no, I want to have new configuration only if the upgrader was successful
otherwise, I'll have a new Config() later, after upgraders

File: src/main/java/com/microfocus/octane/plugins/admin/ConfigResource.java (hunk 1 above)
Comment (pikachugb) [reply src/main/java/com/microfocus/octane/plugins/admin/ConfigResource.java:158]: it can't be null object, only empty string
and if it's anything other than numbers it will throw an error: 
`Workspace id not a valid number: For input string: "sdadas"`

**Review by timisaratean — APPROVED** [review]

**Review by tatizapirtan — APPROVED** [review]

**Review by Rogojan — APPROVED** [review]
